File: Cpp/AdventOfCode/src/2015/d19_RudolfReactor.cpp
#pragma once

#include "2015/d19_RudolfReactor.h"

SOLUTION(2015, 19) {
    using Transforms = Constexpr::SmallMap<std::string, std::vector<std::string>>;
    using RTransforms = Constexpr::SmallMap<std::string, std::string>;

    constexpr void ParseTransform(const std::string & line, Transforms& transforms, RTransforms& reverseTransforms) {
        auto s = Constexpr::Split(line, " => ");
        auto from = std::string(s[0]);
        auto to = std::string(s[1]);
        
        transforms[from].push_back(to);
        reverseTransforms[to] = from;
    }

    constexpr void ExtractMolecules(const std::string& initial, u32 strLen, const Transforms& transforms, Constexpr::BigSet<std::string>& outMolecules) {
        for (size_t i = 0; i < initial.size(); i++) {
            auto candidate = initial.substr(i, strLen);
            if (transforms.contains(candidate)) {
                for (const auto& replacement : transforms.at(candidate)) {
                    auto molecule = initial.substr(0, i) + replacement;
                    if (initial.size() > strLen + i) {
                        molecule += initial.substr(i + strLen);
                    }
                    outMolecules.insert(molecule);
                }
            }
        }
    }

    constexpr bool CollapseMolecule(std::string& molecule, const RTransforms& reverseTransforms) {
        const auto keys = reverseTransforms.GetKeys();
        for (const auto& key : keys) {
            if (molecule.find(key) != molecule.npos) {
                molecule = molecule.replace(molecule.find(key), key.size(), reverseTransforms.at(key));
                return true;
            }
        }

        return false;
    }

    PART_ONE() {
        auto groups = SplitInputIntoGroups(lines);
        Transforms transforms;
        RTransforms reverseTransforms;
        for (const auto& line : groups[0]) {
            ParseTransform(line, transforms, reverseTransforms);
        }

        Constexpr::BigSet<std::string> molecules;
        ExtractMolecules(groups[1][0], 1, transforms, molecules);
        ExtractMolecules(groups[1][0], 2, transforms, molecules);

        return Constexpr::ToString(molecules.size());
    }
    PART_TWO() {
        auto groups = SplitInputIntoGroups(lines);
        Transforms transforms;
        RTransforms reverseTransforms;
        for (const auto& line : groups[0]) {
            ParseTransform(line, transforms, reverseTransforms);
        }

        size_t steps = 0;
        std::string running = groups[1][0];
        while (true) {
            if (!CollapseMolecule(running, reverseTransforms)) break;
            steps++;
        }
        return Constexpr::ToString(steps);
    }
    TESTS() {
        std::vector<std::string> lines = {
            "H => HO",
            "H => OH",
            "O => HH",
            "",
            "HOH"
        };

        if (PartOne(lines) != "4") return false;
        return true;
    }
}

File: Rust/rust_aoc/src/year_2015/d19_rudolf_reactor.rs
use std::collections::{HashMap, HashSet};
use crate::input_utils::split_input_into_groups;

type Transforms<'a> = HashMap<&'a str, Vec<&'a str>>;
type RTransforms<'a> = HashMap<&'a str, &'a str>;

pub fn part_one(lines: &Vec<String>) -> String {
    let groups = split_input_into_groups(&lines);
    let (transforms, _) = parse_transforms(&groups[0]);

    count_molecules(&groups[1][0], &transforms).to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    let groups = split_input_into_groups(&lines);
    let (_, reverse_transforms) = parse_transforms(&groups[0]);
    let mut steps = 0usize;
    let mut running = groups[1][0].clone();

    loop {
        if !collapse_molecule(&mut running, &reverse_transforms) {
            break;
        }
        steps += 1;
    }

    steps.to_string()
}

fn parse_transforms(lines: &Vec<String>) -> (Transforms, RTransforms) {
    let mut transforms = Transforms::new();
    let mut reverses = RTransforms::new();

    for line in lines {
        let s: Vec<&str> = line.split(" => ").collect();
        let from = s[0];
        let to = s[1];

        let values = transforms.entry(from).or_default();
        values.push(to);

        reverses.insert(to, from);
    }

    (transforms, reverses)
}

fn count_molecules(initial: &str, transforms: &Transforms) -> usize {
    let mut molecules: HashSet<String> = HashSet::new();

    for (key, values) in transforms {
        for m in initial.match_indices(key) {
            for v in values {
                let mut next = String::from(initial);
                next.replace_range(m.0 ..m.0 + m.1.len(), v);
                molecules.insert(next);
            }
        }
    }

    molecules.len()
}

fn collapse_molecule(molecule: &mut String, reverses: &RTransforms) -> bool {
    for (key, value) in reverses {
        if let Some(m) = molecule.find(key) {
            molecule.replace_range(m..m + key.len(), value);
            return true;
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part_one_test() {
        let lines: Vec<String> = vec!["H => HO", "H => OH", "O => HH", "", "HOH"].iter().map(|s| String::from(*s)).collect();
        assert_eq!("4", part_one(&lines));
    }
}